man/rump.1
.\" generated with Ronn/v0.7.3
.\" http://github.com/rtomayko/ronn/tree/0.7.3
.
.TH "RUMP" "1" "February 2011" "" ""
.
.SH "NAME"
rump \- Do Puppet runs locally from a Git checkout
.
.SH "SYNOPSIS"
.
.IP "\(bu" 4
\fBrump\fR shake [\fIpuppet\-options\fR]
.
.IP "\(bu" 4
\fBrump\fR clone \fIrepository\fR [directory]
.
.IP "\(bu" 4
\fBrump\fR freeze [\fIproject\fR \fIrepository\fR \-\-release=\fIrelease\fR]
.
.IP "\(bu" 4
\fBrump\fR whoami [rfc2822\-address]
.
.IP "\(bu" 4
\fBrump\fR init \fIproject\fR
.
.IP "\(bu" 4
\fBrump\fR scaffold \fIproject\fR
.
.IP "" 0
.
.SH "DESCRIPTION"
\fBRump\fR helps you run Puppet locally against a Git checkout\.
.
.P
Rump supports a Puppet workflow where you quickly + iteratively develop your Puppet manifests on a single machine, then push your changes up to a repository to deploy to the rest of your infrastructure\.
.
.P
This workflow also complements a Capistrano or MCollective\-style deployment, where you remotely instruct Rump to check out a copy of the latest manifests and run them\.
.
.P
Rump also has the ability to freeze Puppet in to the manifests repository, letting you quickly test different versions of Puppet without waiting for packages to appear, and reducing the dependencies on a system to run Puppet down to just Ruby and git\.
.
.SH "COMMANDS"
.
.TP
\fBclone\fR
Clone a git repository of Puppet manifests\.
.
.IP
\fIrepository\fR is the url of a git repository\. Essentially invokes \fBgit clone\fR\. If [directory] is specified, will clone the repository into the specified directory\.
.
.TP
\fBfreeze\fR
Freeze an arbitrary git repository into the current git repository as a submodule under \fBvendor/\fR\.
.
.IP
With no arguments, \fBfreeze\fR will freeze in Puppet + Facter, to the latest commit on master\. This is especially useful if you want to track the latest development release of Puppet\.
.
.IP
If \fIproject\fR and \fIrepository\fR are specified, a git repository is cloned and set up as a submodule under \fBvendor/\fR\.
.
.IP
If \fB\-\-release\fR is specified, \fBfreeze\fR will create and switch to a branch matching a particular tag or revision\. Use this to lock to a specific version of Puppet\.
.
.IP
Once the repository is frozen in, switch between tagged releases by \fBcd\fRing to the repository under vendor and running \fBgit checkout <tag>\fR\.
.
.IP
Any repository frozen under \fBvendor/\fR will be added to the Puppet load path when using \fBrump shake\fR\.
.
.TP
\fBshake\fR
Do a Puppet run\. Assumes the current directory is a repository of Puppet manifests\.
.
.IP
\fIpuppet\-options\fR are any options you can pass to \fBpuppet\fR on the command line (e\.g\. \fB\-\-debug\fR, \fB\-\-verbose\fR, \fB\-\-tags foobar\fR)\.
.
.TP
\fBhelp\fR
Display the man page\.
.
.TP
\fBinit\fR
Initialise a repo of scaffolded Puppet manifests\. Generates scaffolding for a repository of Puppet manifests, and \fBgit init\fRs the repository\.
.
.TP
\fBscaffold\fR
Generate scaffolding for a repository of Puppet manifests, primed for Rump\.
.
.IP
Called by \fBinit\fR\. You should never have to call this manually\.
.
.TP
\fBwhoami\fR
Sets \fBuser\.name\fR and \fBuser\.email\fR using \fBgit config\fR\. Useful if you have multiple users working with the repository from the same UNIX account\.
.
.IP
If no arguments are specified, will output the currently configured name and email\.
.
.IP
If an argument is specified, treat the argument as an RFC2822 formatted email address, e\.g\. John Doe \fIjohn@example\.org\fR\. Sets \fBuser\.name\fR and \fBuser\.email\fR with \fBgit config\fR\.
.
.IP
Modelled on \fBbzr whoami\fR\.
.
.SH "OPTIONS"
.
.TP
\fB\-\-help|\-h\fR
Display the man page\.
.
.SH "RETURN VALUES"
.
.IP "\(bu" 4
\fB0\fR: Command completed successfully\.
.
.IP "\(bu" 4
\fB1\fR: Argument error\.
.
.IP "\(bu" 4
\fB2\fR: Subcommand error\.
.
.IP "\(bu" 4
\fB3\fR: Required frozen submodule doesn\'t exist\.
.
.IP "" 0
.
.SH "QUIRKS"
.
.IP "1." 4
Puppet\'s fileserver (\fBsource => "puppet:///\.\.\."\fR on File resources) doesn\'t behave as expected on Puppet < 2\.6\. If you are using Puppet < 2\.6, all files need to be templates\.
.
.IP "" 0
.
.SH "EXAMPLES"
Initialise a Rump\-ready Puppet repository:
.
.IP "" 4
.
.nf

$ rump init barfoo
Initialized empty Git repository in barfoo/\.git/
.
.fi
.
.IP "" 0
.
.P
Clone a git repository of Puppet manifests:
.
.IP "" 4
.
.nf

$ rump clone git@github\.com:exampledotorg/manifests\.git
.
.fi
.
.IP "" 0
.
.P
Clone a git repository of Puppet manifests into a specific directory:
.
.IP "" 4
.
.nf

$ rump clone git@github\.com:exampledotorg/manifests\.git another_clone
.
.fi
.
.IP "" 0
.
.P
Graft a Rump scaffold onto an existing repository of manifests:
.
.IP "" 4
.
.nf

$ rump scaffold another_clone
.
.fi
.
.IP "" 0
.
.P
Freeze Puppet and Facter:
.
.IP "" 4
.
.nf

$ rump freeze
Cloning into vendor/puppet\.\.\.
remote: Counting objects: 72007, done\.
remote: Compressing objects: 100% (22120/22120), done\.
remote: Total 72007 (delta 52640), reused 66470 (delta 48322)
Receiving objects: 100% (72007/72007), 12\.13 MiB | 253 KiB/s, done\.
Resolving deltas: 100% (52640/52640), done\.
Cloning into vendor/facter\.\.\.
remote: Counting objects: 4185, done\.
remote: Compressing objects: 100% (1930/1930), done\.
remote: Total 4185 (delta 2202), reused 3973 (delta 2095)
Receiving objects: 100% (4185/4185), 840\.81 KiB | 174 KiB/s, done\.
Resolving deltas: 100% (2202/2202), done\.
Freezing complete\.
.
.fi
.
.IP "" 0
.
.P
Freeze some other repository:
.
.IP "" 4
.
.nf

$ rump freeze moonshine https://github\.com/railsmachine/moonshine\.git
Cloning into vendor/moonshine\.\.\.
remote: Counting objects: 4122, done\.
remote: Compressing objects: 100% (1976/1976), done\.
remote: Total 4122 (delta 2190), reused 3730 (delta 1911)
Receiving objects: 100% (4122/4122), 543\.59 KiB | 58 KiB/s, done\.
Resolving deltas: 100% (2190/2190), done\.
Freezing complete\.
.
.fi
.
.IP "" 0
.
.P
Freeze Puppet at a particular version:
.
.IP "" 4
.
.nf

$ rump freeze puppet git://github\.com/puppetlabs/puppet\.git \-\-release=2\.6\.2
Cloning into /Users/auxesis/foobar/vendor/puppet\.\.\.
remote: Counting objects: 72007, done\.
remote: Compressing objects: 100% (22120/22120), done\.
remote: Total 72007 (delta 52640), reused 66470 (delta 48322)
Receiving objects: 100% (72007/72007), 12\.13 MiB | 238 KiB/s, done\.
Resolving deltas: 100% (52640/52640), done\.
Switched to a new branch \'2\.6\.2\'
Freezing complete\.
.
.fi
.
.IP "" 0
.
.P
Do a Puppet run:
.
.IP "" 4
.
.nf

$ rump shake
You should probably be root when running this! Proceeding anyway\.\.\.
Using frozen Puppet from vendor/puppet\.
notice: Finished catalog run in 0\.01 seconds
.
.fi
.
.IP "" 0
.
.P
Do a Puppet run, and pass in arguments to Puppet:
.
.IP "" 4
.
.nf

$ rump shake \-\-noop \-\-debug
You should probably be root when running this! Proceeding anyway\.\.\.
Using frozen Puppet from vendor/puppet\.
ruby \-I vendor/facter/lib \-I vendor/puppet/lib vendor/puppet/bin/puppet
\-\-modulepath modules \-\-confdir etc \-\-vardir var manifests/site\.pp \-\-noop \-\-debug
debug: Creating default schedules
debug: Failed to load library \'selinux\' for feature \'selinux\'
debug: Puppet::Type::File::ProviderMicrosoft_windows: feature microsoft_windows is missing
debug: Failed to load library \'ldap\' for feature \'ldap\'
debug: Failed to load library \'shadow\' for feature \'libshadow\'
\.\.\.
.
.fi
.
.IP "" 0
.
.P
Find out who the currently configured git user is:
.
.IP "" 4
.
.nf

$ rump whoami
Jane Doe <jane@example\.org>
.
.fi
.
.IP "" 0
.
.P
Set the git user:
.
.IP "" 4
.
.nf

$ rump whoami \'John Doe <john@example\.org>\'
.
.fi
.
.IP "" 0
.
.SH "AUTHOR"
Lindsay Holmwood \fIlindsay@holmwood\.id\.au\fR
.
.SH "COPYRIGHT"
Rump is copyright Rails Machine LLC\.
.
.SH "SEE ALSO"
puppet(8)


lib/rump.rb
#!/usr/bin/env ruby

require 'thor'
require 'pathname'

module Logging
  ESCAPES = { :green  => "\033[0;32m",
              :yellow => "\033[0;33m",
              :red    => "\033[47;31m",
              :reset  => "\033[0m" }

  def info(message)
    emit(:message => message, :color => :green)
  end

  def warn(message)
    emit(:message => message, :color => :yellow)
  end

  def error(message)
    emit(:message => message, :color => :red)
  end

  def emit(opts={})
    color   = opts[:color]
    message = opts[:message]
    print ESCAPES[color]
    print message
    print ESCAPES[:reset]
    print "\n"
  end
end

# Thor's default stack trace on errors is ugly - make it pretty.
class Thor
  class << self
    def handle_argument_error(task, error)
      puts "#{task.name.inspect} was called incorrectly. Call as #{task.formatted_usage(self, banner_base == "thor").inspect}."
      exit 1
    end
  end
end


class Rump < Thor
  include Logging

  def initialize
    super
    @root = Pathname.new(Dir.getwd)
    @install_root = Pathname.new(File.expand_path(File.join(File.dirname(__FILE__))))
  end

  desc "clone <repository> [directory]", "clone a Git repository of Puppet manifests"
  def clone(repository, directory=nil)
    abort_unless_git_installed
    directory ||= File.basename(repository.split('/').last, '.git')
    command = "git clone -q #{repository} #{directory}"
    system(command)

    # A cloned repo may have submodules - automatically initialise them.
    if File.exists?(File.join(directory, '.gitmodules'))
      Dir.chdir(directory) do
        system("git submodule init")
        system("git submodule update")
      end
    end
  end

  desc "shake [puppet arguments]", "do a local Puppet run"
  def shake(*puppet_args)
    if ENV['USER'] != "root"
      warn "You should probably be root when running this! Proceeding anyway..."
    end

    args = []
    if has_frozen_components?
      args << "ruby"
      Dir.glob("#{@root.join('vendor')}/*").each do |dir|
        args << "-I #{@root.join('vendor', dir, 'lib')}"
      end
      args << "#{@root.join('vendor', 'puppet', 'bin', 'puppet')}"
      info "Using frozen Puppet from #{@root.join('vendor', 'puppet')}."
    else
      abort_unless_puppet_installed(:message => "Please either install it on your system or freeze it with 'rump freeze'")
      args << "puppet"
    end
    args << "--modulepath #{@root.join('modules')}"
    args << "--confdir #{@root.join('etc')}" unless puppet_args.include?("--confdir")
    args << "--vardir #{@root.join('var')}" unless puppet_args.include?("--vardir")
    args << "#{@root.join('manifests', 'site.pp')}"

    args += puppet_args

    command = args.join(' ')
    puts command if args.include?("--debug")
    system(command) ? exit(0) : exit(2)
  end

  desc "freeze [repository, project]", "freeze Puppet into your manifests repository"
  def freeze(*args)
    abort_unless_git_installed

    commands = []
    if args.size >= 2
      project    = args[0]
      repository = args[1]
      commands << { :command => "git submodule add #{repository} #{@root.join('vendor', project)}" }
      if args.detect { |arg| arg =~ /^--release\=(.+)/ }
        version = $1
        commands << { :command => "git checkout -b #{version} #{version}", :directory => @root.join('vendor', project) }
      end
      commands << { :command => "git commit --quiet -am 'Froze in #{repository} as a submodule'" }
    else
      commands << { :command => "git submodule add git://github.com/puppetlabs/puppet.git #{@root.join('vendor', 'puppet')}" }
      commands << { :command => "git submodule add git://github.com/puppetlabs/facter.git #{@root.join('vendor', 'facter')}" }
      commands << { :command => "git commit --quiet -am 'Froze in Puppet + Facter as submodules'" }
    end

    commands.each do |attrs|
      dir = attrs[:directory] || @root
      Dir.chdir(dir) do
        exit(2) unless system(attrs[:command])
      end
    end

    info "Freezing complete."
  end

  desc "scaffold <project>", "generate scaffolding for a repository of Puppet manifests"
  def scaffold(project)
    [ @root + project,
      @root + project + 'manifests',
      @root + project + 'modules',
      @root + project + 'var/reports',
      @root + project + 'vendor' ].each do |directory|
      FileUtils.mkdir_p(directory)
    end

    File.open(@root.join(project, 'README.md'), 'w') do |f|
      f << <<-README.gsub(/^ {8}/, '')
        #{project} manifests
        #{"=" * project.size}==========

        modules/ <= Puppet modules
        manifests/ <= Puppet nodes
        vendor/ <= frozen Puppet + Facter

        Running Puppet with Rump
        ------------------------

        From within this directory, run:

            rump shake

        You can pass options to Puppet after the 'shake':

            rump shake --debug --test

        Freezing Puppet
        ---------------

        Firstly, you need to create a git repository:

            git init

        Now you can freeze Puppet:

            rump freeze

        Once Rump has frozen Puppet, commit the changes:

            git commit -m 'added facter + puppet submodules' .

        Now Rump will use the frozen Puppet when you run 'rump shake'.

      README
    end
  end

  desc "init <project>", "initialise a repo of scaffolded Puppet manifests"
  def init(project)
    scaffold(project)
    repo_path = @root.join(project)
    template_path = @install_root.join('generators', 'git')

    Dir.chdir(repo_path) do
      commands = [ "git init --quiet --template=#{template_path}",
                   "git add .",
                   "git commit --quiet -am 'Initial commit.'" ]
      commands.each do |command|
        system(command)
      end
    end

    info "Your new Rump project has been initialised in #{repo_path}"
  end

  desc "whoami [rfc2822-address]", "set the current commit author"
  def whoami(address=nil)
    # getter
    if address
      name  = address[/^(.+)\s+</, 1]
      email = address[/<(.+)>$/, 1]

      unless name && email
        abort("Supplied address isn't a valid rfc2822 email address")
      end

      system("git config user.name '#{name}'")
      system("git config user.email '#{email}'")
    # setter
    else
      name = `git config user.name`.strip
      email = `git config user.email`.strip

      if name.empty? || email.empty?
        warn "You don't have a name or email set."
      else
        puts "#{name} <#{email}>"
      end
    end
  end

  private
  def has_frozen_components?
    vendored = Dir.glob("#{@root.join('vendor')}/*").map {|v| v.split('/').last}
    vendored.include?("puppet") && vendored.include?("facter")
  end

  # helper + abortive methods
  %w(puppet git).each do |bin|
    class_eval <<-METHOD, __FILE__, __LINE__
      no_tasks do
        def #{bin}_installed?
          `which #{bin}` =~ /#{bin}$/ ? true : false
        end

        def abort_unless_#{bin}_installed(opts={})
          unless #{bin}_installed?
            error "You don't have #{bin.capitalize} installed!"
            error opts[:message] || "Please install it on your system."
            exit 3
          end
        end
      end
    METHOD
  end

end

